Use separate topic for lifecycle transition tasks

welansari opened the pull request:
The lifecycle conductor uses the lag on the object tasks to throttle the lifecycle scan.
However, the object tasks topic handles events that are not in the direct path and that does not impact the lifecycle scan.
We use a different topic for those events to avoid unwanted throttle.

Issue: BB-467

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-467 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.30`

* `8.7.0`


Please check the `Fix Version/s` of BB-467, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

Commit:
use separate topic for lifecycle transition tasks

The lifecycle conductor uses the lag on the object tasks topic to throttle the lifecycle scan.
However, the object tasks topic handles events that are not in the direct path and that
does not impact the lifecycle scan.
We use a different topic for those events to avoid unwanted throttle.

Issue: BB-467

Commit:
adapt lifecycle tests to support transition topic

Issue: BB-467

francoisferrand commented on extensions/lifecycle/LifecycleConfigValidator.js:
```diff
@@ -12,6 +12,7 @@ const joiSchema = joi.object({
     zookeeperPath: joi.string().required(),
     bucketTasksTopic: joi.string().required(),
     objectTasksTopic: joi.string().required(),
+    transitionTasksTopic: joi.string().required(),
```
should it be required (i.e. require ZKOP to set the value on newer backbeat versions), or can we default to objectTasksTopic if this field is not specified (i.e. maintain backwards compatibility, degrading to the old behavior) ?

This may be done like this:

```suggest
transitionTasksTopic: joi.string().required().default(config => config.objectTaskTopic),
```

francoisferrand commented on extensions/lifecycle/objectProcessor/LifecycleObjectProcessor.js:
```diff
@@ -94,9 +93,11 @@ class LifecycleObjectProcessor extends EventEmitter {
     }
 
-    getConsumerParams() {
+    getConsumerParams(isTransitionProcessor = false) {
```
instead of adding the topic here, it should be added/overriden directly in the derived class: e.g. in `transitionProcessor.getConsumerParams()`

francoisferrand commented on tests/functional/lifecycle/LifecycleTask.js:
```diff
@@ -328,10 +334,10 @@ class ProducerMock {
             this.sendCount.bucket++;
             this.entries.bucket.push(entry);
-        } else if (topicName === 'object-tasks') {
+        } else if (topicName === objectTasksTopic) {
```
shoudl we also count entries pushed to transition topic ?

welansari replied:
No need as nothing is pushed into it here, only ones pushing to it are either sorbet or the replication process

francoisferrand commented on extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js:
```diff
@@ -48,8 +49,32 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
         return 'transition-processor';
     }
 
+    _getTransitionTaskConsumerParams() {
```
should be refactored to avoid duplicating the actual "definition" of the params

maybe introduce a function `getConsumerParam(topic)` in ObjectProcessor, or just add an (optional) param to `getConsumerParams(topic = this._lcConfig.objecTasksTopic)`...

francoisferrand commented on extensions/lifecycle/objectProcessor/LifecycleObjectProcessor.js:
```diff
@@ -73,7 +72,7 @@ class LifecycleObjectProcessor extends EventEmitter {
     }
 
-    _getObjecTaskConsumerParams() {
+    _getObjecTaskConsumerParams(topic) {
```
nit: `_getTopicConsumerParams` ?

KillianG commented on tests/unit/lifecycle/LifecycleObjectExpirationProcessor.spec.js:
```diff
@@ -0,0 +1,26 @@
+        );
+    });
+
+    it('consumer params should contain object tasks topic', () => {
```
```suggestion
    it('should contain object tasks topic in consumer params', () => {
```

KillianG commented on tests/unit/lifecycle/LifecycleObjectTransitionProcessor.spec.js:
```diff
@@ -0,0 +1,26 @@
+        );
+    });
+
+    it('consumer params should contain transition tasks topic', () => {
```
```suggestion
    it('should contain transition tasks topic in consumer params', () => {
```

KillianG reviewed (approved):
2 little comment about phrasing the tests

welansari commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-467-separate-transition-topic` into
`development/8.6`
* [w/8.7/bugfix/BB-467-separate-transition-topic](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-467-separate-transition-topic) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/backbeat/commit/b78151bd2153f089fe9c9783945df3c2936b70a2) did not succeed in branch w/8.7/bugfix/BB-467-separate-transition-topic.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-467.

Goodbye kerkesni.
